static/js/util.ts

```diff
@@ -47,34 +47,48 @@ export const lower_same = function lower_same(a: string, b: string): boolean {
     return a.toLowerCase() === b.toLowerCase();
 };
 
+export type StreamTopic = {
+    stream_id: number;
+    topic: string;
+};
+
 export const same_stream_and_topic = function util_same_stream_and_topic(
-    a: Message,
-    b: Message,
+    a: StreamTopic,
+    b: StreamTopic,
 ): boolean {
     // Streams and topics are case-insensitive.
     return a.stream_id === b.stream_id && lower_same(a.topic, b.topic);
 };
 
-export function is_pm_recipient(user_id: number, message: Message): boolean {
-    const recipients = message.to_user_ids.split(",");
+export function is_pm_recipient(user_id: number, message_recipient: PMRecipient): boolean {
+    const recipients = message_recipient.to_user_ids.split(",");
     return recipients.includes(user_id.toString());
 }
 
 export function extract_pm_recipients(recipients: string): string[] {
     return recipients.split(/\s*[,;]\s*/).filter((recipient) => recipient.trim() !== "");
 }
 
-export const same_recipient = function util_same_recipient(a: Message, b: Message): boolean {
+// When the type is "private", properties from PMRecipient, namely to_user_ids might be
+// undefined.
+type Recipient =
+    | {type: string}
+    | ((Partial<PMRecipient> & {type: "private"}) | (StreamTopic & {type: "stream"}));
+
+export const same_recipient = function util_same_recipient(a?: Recipient, b?: Recipient): boolean {
     if (a === undefined || b === undefined) {
         return false;
     }
 
-    if (a.type === "private" && b.type === "private") {
+    if (a.type === "private" && b.type === "private" && "to_user_ids" in a && "to_user_ids" in b) {
         if (a.to_user_ids === undefined) {
             return false;
         }
         return a.to_user_ids === b.to_user_ids;
-    } else if (a.type === "stream" && b.type === "stream") {
+    }
+    // For type narrowing, checking the presence of "stream_id" is enough to let TypeScript know that
+    // the StreamTopic is a subtype of the recipient's type.
```

Comment:
If `Recipient` didn’t have that catch-all `{type: string}` alternative, `a.type === "stream"` would be enough.

Reply:
That's true, but since `util_same_recipient` also accepts objects like `{type: "unknown type"}`, the catch-all is needed. I think the best type here should be `unknown`, but type narrowing with `in` is not supported.

Reply:
All of our real `Recipient` objects will be `private` or `stream`; we only use `{type: "unknown type"}` in an overly-defensive test. We should delete that test. We want to use the type system to help enforce this invariant, so that we don’t *need* overly-defensive tests.

(We should eventually have defensive code, but it should go at the I/O boundary between our typed code and the untyped external world, not scattered in random internal functions. [Zod](https://zod.dev/) is an excellent TypeScript library for this.)

Reply:
OK, that makes sense.

Reply:
Having corrected the type, we can remove this chunk of changes to `util_same_recipient`.

Reply:
Fixed.